Fix communication race bug in interpolation framework

knelli2 opened the pull request:
## Proposed changes

We were not checking the order we received temporal IDs before sending points to the interpolator and this made some horizon finds happen out of order. I don't think this had a significant impact, but still best to fix it. In order to make some of the logic easier, I did a couple things

- Explicitly only allow sequential targets in the interpolator. This was already implicitly done because horizon finds are sequential and no other interpolation used the Interpolator. But best to make it explicit. (commit 2)
- There were multiple times I found where an action running on the target singleton would schedule another action on itself when it could just call it directly. So I did that. (commit 3)

I tested this branch on a BBH test case that I know to fail, and it has gotten significantly past the point of failure

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Do less communication on interpolation target

knelli2 committed:
Only allow sequential targets in interpolator

knelli2 committed:
Fix includes in testing framework

knelli2 committed:
Fix comm race bug in interpolation framework

wthrowe commented on src/ParallelAlgorithms/Interpolation/Actions/AddTemporalIdsToInterpolationTarget.hpp:
```diff
@@ -54,56 +54,40 @@ struct AddTemporalIdsToInterpolationTarget {
                     Parallel::GlobalCache<Metavariables>& cache,
```
Update this class's documentation.

wthrowe commented on src/ParallelAlgorithms/Interpolation/Actions/VerifyTemporalIdsAndSendPoints.hpp:
```diff
@@ -40,23 +47,6 @@ void verify_temporal_ids_and_send_points_time_independent(
-        // order. Note that this isn't a permanent actual fix, but so far works
-        // in practice.
-        alg::sort(*ids);
-        alg::sort(*pending_ids);
```
Can you confirm that the changes in this commit fix this issue?  The comment here says it's a race related to receiving messages from the elements, which doesn't sound related to the changes here.  Same thing again later.

Actually, this is probably just in the wrong commit.

wthrowe commented on tests/Unit/ParallelAlgorithms/Interpolation/Test_AddTemporalIdsToInterpolationTarget.cpp:
```diff
@@ -150,12 +183,19 @@ void test_add_temporal_ids() {
   ActionTesting::set_phase(make_not_null(&runner), Parallel::Phase::Testing);
+  auto& target_box =
+      ActionTesting::get_databox<target_component>(make_not_null(&runner), 0);
+  (void)target_box;
```
Why this line?  It appears to be used unconditionally later.

wthrowe commented on src/ParallelAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -411,18 +411,94 @@ void flag_temporal_id_as_pending(const gsl::not_null<db::DataBox<DbTags>*> box,
+                Parallel::printf(
+                    "%s: Adding %s to %s\n"
+                    " Pending: %s\n"
+                    " Pending pending: %s\n",
```
This seems like it should be controlled by some verbosity setting.

wthrowe commented on src/ParallelAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -411,18 +411,94 @@ void flag_temporal_id_as_pending(const gsl::not_null<db::DataBox<DbTags>*> box,
+              }
+            }
+
+            alg::sort(*pending_ids);
```
This one should always be in order, no?

wthrowe commented on src/ParallelAlgorithms/Interpolation/Actions/InterpolationTargetReceiveVars.hpp:
```diff
@@ -182,6 +185,38 @@ struct InterpolationTargetReceiveVars {
+                      most_recent_completed_id.id) {
+                    Parallel::printf(
+                        "%s: Moving %s from pending pending to pending and "
+                        "sending points\n",
```
Again, this should probably have some verbosity control.

wthrowe commented on src/ParallelAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -411,18 +411,94 @@ void flag_temporal_id_as_pending(const gsl::not_null<db::DataBox<DbTags>*> box,
+                      print_pending("pending pending (pending not empty)");
+                    }
+                  };
+              // Cases: (1p = pending, 2p = pending pending)
```
[Arbitrary line] This seems overcomplicated.  Instead of multiple levels of pending, could you just make the pending list a priority queue and check if the first element's previous link is pointing to the last completed id before running it?

knelli2 replied:
Hmm possibly. I think in order to make this change, the bulk of this logic would move into `VerifyTemporalIdsAndSendPoints` because there it loops over pending ids to see if they are ready, and then this function would basically remain unchanged (except now the sort of pending ids would always be needed to ensure the earliest id is first).

I also think with all the changes in this PR that there can only be a single id in `Tags::TemporalIds` at a time (since only one horizon find can be happening at once). That makes more things easier I think.

wthrowe replied:
If you're not familiar with the [priority queue](https://en.cppreference.com/w/cpp/container/priority_queue) I suggest you read up on that data structure.

knelli2 replied:
I looked over the priority queue and found that it won't be suitable for our case. Within in the source code (and tests) we need to search through the queue to see if an id is already there and the priority queue does not offer iterators or really any way to query what is in the queue besides the `top` value and size. Therefore, I'm just going to stick with a `deque` and sort it.

wthrowe commented on src/ParallelAlgorithms/Interpolation/Actions/VerifyTemporalIdsAndSendPoints.hpp:
```diff
@@ -266,41 +84,142 @@ struct VerifyTemporalIdsAndSendPoints {
+                                               &this_proxy, &next_pending_id,
+                                               &min_expiration_time]() {
+          if constexpr (Parallel::is_in_mutable_global_cache<
+                            Metavariables, domain::Tags::FunctionsOfTime>) {
```
Under what circumstances is FunctionsOfTime in the cache but not the mutable cache?

knelli2 replied:
Oops, this was just copied from one of the other functions I removed. This `if constexpr` can be removed I believe

wthrowe commented on tests/Unit/ParallelAlgorithms/Interpolation/Test_AddTemporalIdsToInterpolationTarget.cpp:
```diff
@@ -312,6 +348,216 @@ void test_add_temporal_ids() {
+
+  const auto insert_id_into = [&]<template <typename> typename Tag>(
+                                  const LinkedMessageId<double>& id,
+                                  const bool insert_at_back = true) {
```
Last argument isn't used.

wthrowe reviewed (changes requested):
Yes, this seems cleaner.

knelli2 commented:
@wthrowe I implemented your suggestions (except for the priority queue), updated dox, and changed the current temporal id tag type to optional if the target is sequential (since only one interpolation can happen at a time for sequential targets). I think this should be good to go. I tested the first fixup and that worked on a BBH case that was failing before and I'm testing the second fixup now.

wthrowe commented on src/ParallelAlgorithms/Interpolation/Actions/AddTemporalIdsToInterpolationTarget.hpp:
```diff
@@ -18,36 +18,33 @@
+///    - CurrentTemporalId doesn't have a value
+///    - PendingTemporalIds is not empty,
+///    - The temporal id is a LinkedMessageId or PendingTemporalIds was empty
+///      upon entry
```
Are these conditions ORed or ANDed?

wthrowe reviewed (changes requested):
OK.  Go ahead and squash once you're satisfied with your testing.

knelli2 commented:
Second fixup also passed my test. Squashed everything in

knelli2 commented:
@nilsdeppe Do you want to look at this? I believe it will fix the issue you ran into testing nodegroups

nilsdeppe commented:
I'll take a quick glance today, but won't do a detailed review since Will did that :)
